src/GNDStk/gnds-utility.hpp

```diff
@@ -0,0 +1,231 @@
+
+// convert
+// ...Tree to Tree
+// ...Tree to xml
+// ...Tree to json
+// ...xml  to Tree
+// ...xml  to xml
+// ...xml  to json
+// ...json to Tree
+// ...json to xml
+// ...json to json
+
+
+// Tree to {Tree,xml,json}
+template<
+   template<class...> class MCONFROM,
+   template<class...> class CCONFROM,
+   template<class...> class MCONTO,
+   template<class...> class CCONTO
+>
+bool convert(
+   const gnds::Tree<MCONFROM,CCONFROM> &,
+   gnds::Tree<MCONTO,CCONTO> &
+);
+
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+bool convert(
+   const gnds::Tree<MCON,CCON> &,
+   gnds::xml &
+);
+
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+bool convert(
+   const gnds::Tree<MCON,CCON> &,
+   gnds::json &
+);
+
+
+// xml to {Tree,xml,json}
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+bool convert(const gnds::xml &, gnds::Tree<MCON,CCON> &);
+
+bool convert(const gnds::xml &, gnds::xml &);
+
+bool convert(const gnds::xml &, gnds::json &);
+
+
+// json to {Tree,xml,json}
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+bool convert(const gnds::json &, gnds::Tree<MCON,CCON> &);
+
+bool convert(const gnds::json &, gnds::xml &);
+
+bool convert(const gnds::json &, gnds::json &);
+
+
+
+// -----------------------------------------------------------------------------
+// Some string constants
+// -----------------------------------------------------------------------------
+
+namespace detail {
+
+// For pugi::xml_node_types { node_pcdata, node_cdata, node_comment }
+inline const std::string keyword_pcdata  = "body";
+inline const std::string keyword_cdata   = "text";
+inline const std::string keyword_comment = "comment";
+
+// Aliases
+inline const std::string &keyword_body = keyword_pcdata;
+inline const std::string &keyword_text = keyword_cdata;
+
+} // namespace detail
+
+
+
+// -----------------------------------------------------------------------------
+// Utility constructs
+// The functions here could possibly go into the detail namespace, but could
+// arguably be useful, in their own right, to users. So, I'll leave them out
+// in the overall project namespace (which enclosed the #include of this file).
+// -----------------------------------------------------------------------------
+
+// filesize(char *)
+inline std::ifstream::pos_type filesize(const char * const file)
+{
+   std::ifstream ifs(file, std::ifstream::ate | std::ifstream::binary);
+   return ifs.tellg();
+}
+
+// filesize(string)
+inline std::ifstream::pos_type filesize(const std::string &file)
+{
+   return filesize(file.c_str());
+}
+
+
+// endsin
+// C++20 will have an ends_with(); for now we'll have this
+inline bool endsin(const std::string &str, const std::string &end)
+{
+   return str.size() >= end.size() && &str[str.size()-end.size()] == end;
+}
+
+// typeof
+#ifdef GNDS_BOOST
```

Comment:
Please remove this.

Reply:
OK, I've removed it. With some of the template metaprogramming that'll be coming down the pipe, some of my own separate tests (and, well, just the work I'm doing to get things right) involves inspecting what *types* pop out of certain complicated constructs. Bringing in a particular boost library conditionally (conditionally so that I must switch it on explicitly, and it isn't imposed on anyone else) was a convenient way of getting what I needed. But I can do it in my own tests by just bringing in what I need separately from GNDStk.